File: ios/auto-mobile-sdk/Sources/AutoMobileSDK/Context/SdkContext.swift
import Foundation

/// Thread-safe mutable context holding ambient state attached to SDK events.
final class SdkContext: @unchecked Sendable {
    private let lock = NSLock()
    private var _sessionId: String?
    private var _userId: String?
    private var _appVersion: String?
    private var _tags: [String: String] = [:]

    init() {}

    var sessionId: String? {
        get { lock.lock(); defer { lock.unlock() }; return _sessionId }
        set { lock.lock(); _sessionId = newValue; lock.unlock() }
    }

    var userId: String? {
        get { lock.lock(); defer { lock.unlock() }; return _userId }
        set { lock.lock(); _userId = newValue; lock.unlock() }
    }

    var appVersion: String? {
        get { lock.lock(); defer { lock.unlock() }; return _appVersion }
        set { lock.lock(); _appVersion = newValue; lock.unlock() }
    }

    func setTag(_ key: String, value: String) {
        lock.lock(); _tags[key] = value; lock.unlock()
    }

    func removeTag(_ key: String) {
        lock.lock(); _tags.removeValue(forKey: key); lock.unlock()
    }

    func clearTags() {
        lock.lock(); _tags.removeAll(); lock.unlock()
    }

    /// Returns an immutable snapshot.
    func snapshot() -> SdkContextSnapshot {
        lock.lock()
        defer { lock.unlock() }
        return SdkContextSnapshot(
            sessionId: _sessionId, userId: _userId,
            appVersion: _appVersion, tags: _tags
        )
    }

    func reset() {
        lock.lock()
        _sessionId = nil; _userId = nil; _appVersion = nil; _tags.removeAll()
        lock.unlock()
    }
}

/// Immutable snapshot of SDK context.
public struct SdkContextSnapshot: Codable, Sendable, Equatable {
    public let sessionId: String?
    public let userId: String?
    public let appVersion: String?
    public let tags: [String: String]
}

File: android/auto-mobile-sdk/src/main/kotlin/dev/jasonpearson/automobile/sdk/context/SdkContext.kt
package dev.jasonpearson.automobile.sdk.context

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/** Thread-safe mutable context holding ambient state attached to SDK events. */
internal class SdkContext {
  private val lock = ReentrantLock()

  @Volatile var sessionId: String? = null
  @Volatile var userId: String? = null
  @Volatile var tenantId: String? = null
  @Volatile var currentScreen: String? = null
  @Volatile var appVersion: String? = null

  private val _tags = mutableMapOf<String, String>()

  fun setTag(key: String, value: String) {
    lock.withLock { _tags[key] = value }
  }

  fun removeTag(key: String) {
    lock.withLock { _tags.remove(key) }
  }

  fun clearTags() {
    lock.withLock { _tags.clear() }
  }

  /** Returns an immutable snapshot of the current context. */
  fun snapshot(): SdkContextSnapshot {
    lock.withLock {
      return SdkContextSnapshot(
        sessionId = sessionId,
        userId = userId,
        tenantId = tenantId,
        currentScreen = currentScreen,
        appVersion = appVersion,
        tags = HashMap(_tags),
      )
    }
  }

  /** Reset all context to defaults. */
  fun reset() {
    lock.withLock {
      sessionId = null
      userId = null
      tenantId = null
      currentScreen = null
      appVersion = null
      _tags.clear()
    }
  }
}

/** Immutable snapshot of SDK context at a point in time. */
data class SdkContextSnapshot(
  val sessionId: String?,
  val userId: String?,
  val appVersion: String?,
  val tags: Map<String, String>,
  val tenantId: String? = null,
  val currentScreen: String? = null,
) {
  /** Preserves the pre-context-extension constructor for binary compatibility. */
  constructor(
    sessionId: String?,
    userId: String?,
    appVersion: String?,
    tags: Map<String, String>,
  ) : this(sessionId, userId, appVersion, tags, null, null)
}
